### PR TITLE
tenant-offload: delete frozen tenants in cloud on tenant or class deletion

### PR DESCRIPTION
**Pull request by moogacs**
### What's being changed:
with this PR change in case of `tenant` or `class` deletion request and the offload module is enabled it will delete the frozen tenants 

this PR adds
-  functionality 
- metric for deletion 
- acceptance test

### Links 
 - https://github.com/weaviate/weaviate-issues/issues/22
### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/migrator.go [hunk 1]
```diff
@@ -128,7 +128,14 @@ func (m *Migrator) AddClass(ctx context.Context, class *models.Class,
 }
 
 func (m *Migrator) DropClass(ctx context.Context, className string) error {
-	return m.db.DeleteIndex(schema.ClassName(className))
+	if err := m.db.DeleteIndex(schema.ClassName(className)); err != nil {
+		return err
+	}
+
+	if m.cloud != nil {
```
Comment (reyreaud-l) [inline adapters/repos/db/migrator.go:135]: If we error deleting the index in the DB (for whichever reason) do we want to *not* delete the cloud shard(s) ? Or should we ignore the error above and still complete the deletion of the cloud shard as well ?

**Review by reyreaud-l — APPROVED** [review]

File: adapters/repos/db/migrator.go (hunk 1 above)
Comment (moogacs) [reply adapters/repos/db/migrator.go:135]: i'd say if it wasn't deleted from the db it shouldn't be deleted on the cloud

File: adapters/repos/db/index.go [hunk 2]
```diff
@@ -1987,6 +1988,28 @@ func (i *Index) dropShards(names []string) error {
 	return ec.ToError()
 }
 
+func (i *Index) dropCloudShards(ctx context.Context, cloud modulecapabilities.OffloadCloud, names []string, nodeId string) error {
+	i.shardTransferMutex.RLock()
+	defer i.shardTransferMutex.RUnlock()
+
+	ec := &errorcompounder.ErrorCompounder{}
```
Comment (antas-marcin) [inline adapters/repos/db/index.go:1995]: I think that `ec` will always return nil bc you just instantiate it here and then you do nothing with it. I think that this needs to removed and instead you should do `return eg.Wait()`

Comment (moogacs) [reply adapters/repos/db/index.go:1995]: good catch, actually should add the error from deletion to it

Comment (moogacs) [reply adapters/repos/db/index.go:1995]: will open a follow up PR

File: modules/offload-s3/module.go [hunk 3]
```diff
@@ -294,16 +302,43 @@ func (m *Module) Download(ctx context.Context, className, shardName, nodeName st
...
+	if nodeName == "" && shardName != "" {
+		return fmt.Errorf("can't pass empty node name")
+	}
+
 	ctx, cancel := context.WithTimeout(ctx, m.timeout)
 	defer cancel()
+
+	cloudPath := fmt.Sprintf("s3://%s/%s/%s/%s/*", m.Bucket, strings.ToLower(className), shardName, nodeName)
+
+	if shardName == "" && nodeName == "" {
```
Comment (antas-marcin) [inline modules/offload-s3/module.go:325]: hmm doesn't previous checks (line 312 and line 316) ensure enough that we here always will have not empty shard and node names?

File: adapters/repos/db/index.go (hunk 2 above)
Comment (moogacs) [reply adapters/repos/db/index.go:1995]: https://github.com/weaviate/weaviate/pull/5303

File: modules/offload-s3/module.go (hunk 3 above)
Comment (moogacs) [reply modules/offload-s3/module.go:325]: here actually it's not validation , it's updating the path to make sure to delete all the shards under a class

**Pull request merged**